--- a/networking/udhcp/dhcpc.c
+++ b/networking/udhcp/dhcpc.c
@@ -939,7 +939,7 @@ static void client_background(void)
 int udhcpc_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int udhcpc_main(int argc UNUSED_PARAM, char **argv)
 {
-	uint8_t *temp, *message;
+	uint8_t *message;
 	const char *str_V, *str_h, *str_F, *str_r;
 	IF_FEATURE_UDHCP_PORT(char *str_P;)
 	void *clientid_mac_ptr;
@@ -1379,6 +1379,7 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 		case INIT_SELECTING:
 			/* Must be a DHCPOFFER to one of our xid's */
 			if (*message == DHCPOFFER) {
+				uint8_t *temp;
 		/* TODO: why we don't just fetch server's IP from IP header? */
 				temp = udhcp_get_option(&packet, DHCP_SERVER_ID);
 				if (!temp) {
@@ -1403,6 +1404,7 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 		case RENEW_REQUESTED:
 		case REBINDING:
 			if (*message == DHCPACK) {
+				uint8_t *temp;
 				temp = udhcp_get_option(&packet, DHCP_LEASE_TIME);
 				if (!temp) {
 					bb_error_msg("no lease time with ACK, using 1 hour lease");
@@ -1480,6 +1482,27 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 				continue; /* back to main loop */
 			}
 			if (*message == DHCPNAK) {
+				/* If network has more than one DHCP server,
+				 * "wrong" server can reply first, with a NAK.
+				 * Do not interpret it as a NAK from "our" server.
+				 */
+				if (server_addr != 0) {
+					uint32_t svid;
+					uint8_t *temp;
+	
+					temp = udhcp_get_option(&packet, DHCP_SERVER_ID);
+					if (!temp) {
+						non_matching_svid:
+						log1("%s with wrong server ID, ignoring packet",
+								"Received DHCP NAK"
+							);
+						continue;
+					}
+					move_from_unaligned32(svid, temp);
+					if (svid != server_addr)
+						goto non_matching_svid;
+			}
+
 				/* return to init state */
 				bb_info_msg("Received DHCP NAK");
 				udhcp_run_script(&packet, "nak");
