--- www/lib/onsen/js/onsenui.js.orig	2015-04-28 10:39:26.323141352 +0200
+++ www/lib/onsen/js/onsenui.js	2015-04-28 13:51:58.861135299 +0200
@@ -21882,17 +21884,19 @@
 
       if (buttonLabels.length <= 2) {
         buttonEl.addClass('alert-dialog-button--one');
       }
 
-      buttonEl.on('click', function() {
+      buttonEl.on('click', function(evt) {
         buttonEl.off('click');
 
         alertDialog.hide({
           callback: function() {
             if (promptDialog) {
-              callback(inputEl.val());
+              if (evt.target.className.indexOf('alert-dialog-button--primal') !== -1 ) {
+                callback(inputEl.val());
+              }
             } else {
               callback(i);
             }
             alertDialog.destroy();
             alertDialog = inputEl = buttonEl = null;
@@ -22023,11 +22027,12 @@
      * @param {Function} [options.callback]
      * @param {Boolean} [options.autofocus]
      */
     prompt: function(options) {
       var defaults = {
-        buttonLabel: 'OK',
+        buttonLabels: ['OK'],
+        primaryButtonIndex: 0,
         animation: 'default',
         title: 'Alert',
         placeholder: '',
         callback: function() {},
         cancelable: false,
@@ -22040,12 +22045,12 @@
       }
 
       createAlertDialog(
         options.title,
         options.message || options.messageHTML,
-        [options.buttonLabel],
-        0,
+        options.buttonLabels,
+        options.primaryButtonIndex,
         options.modifier,
         options.animation,
         options.callback,
         !options.message ? true : false,
         options.cancelable,
